src/util.jl

```diff
@@ -14,20 +14,21 @@ end
 
 # Base gives up on tuples for promote_eltype... (TODO can we improve Base?)
 @generated function promote_tuple_eltype{T <: Tuple}(::Union{T,Type{T}})
-    t = Union{}
+    types = []
     for i = 1:length(T.parameters)
         tmp = T.parameters[i]
         if tmp <: Vararg
             tmp = tmp.parameters[1]
         end
-        t = promote_type(t, tmp)
+        push!(types, tmp)
     end
     return quote
         $(Expr(:meta,:pure))
```

Comment:
NB: I don't actually know what `Expr(:meta, :pure))` does here, so I left it alone